### ISSUE TITLE
Fully-Fused MLP implementation

### ISSUE THREAD

**Issue by salykovaa** [google/jax#15823]
Dear all,

Developers from Nvidia have recently fused MLP network in a large mega kernel and called it "Fully-Fused MLP", which is 5x-10x faster than the equivalent MLP implementation from PyTorch or TensorFlow. People currently use these fully-fused MLPs to accelerate training and inference of ML applications, particularly they use them extensively in neural computer graphics to obtain high gains in speed. I would like to know do you maybe plan to implement it or do you have an equivalent implementation? (Im not sure if this can be achieved by simply applying `jax.jit` on normal MLP).

I think it is a nice feature, which will enrich the jax library and make it more attractive for machine learning / computer graphics researchers and engineers.

<img width="847" alt="image" src="https://user-images.githubusercontent.com/63703454/229091825-8126ac7f-7c1e-4e46-b7dc-599696b5d030.png">

References:
[1] [Real-time Neural Radiance Caching for Path Tracing](https://arxiv.org/abs/2106.12372)
[2] [Technical Paper](https://www.nvidia.com/en-us/on-demand/session/gtcspring21-e31307/)
[3] [Nvidia's Fully fused MLP implementation](https://github.com/NVlabs/tiny-cuda-nn)

**Comment by terafo** [comment]
The advantages of this approach mainly apply to smaller networks with widths up to 128 and completely disappear with MLPs that have width 256 and higher. Initially, the training advantage was about 2x, but this gap has only diminished over time. Though it might be interesting to pinpoint the precise performance difference, developing a custom fully fused MLP in JAX appears unwarranted. The main reason is that the method is highly specialized and the benefits are not substantial enough. Furthermore, attempting such an integration would conflict with JAX's core values such as flexibility and simplicity. Introducing a highly specialized solution would undermine these fundamental principles.

**Comment by salykovaa** [comment]
@terafo 

Thanks for your answer!
